cirq-core/cirq/ops/hamiltonian_gate.py

```diff
@@ -0,0 +1,354 @@
+
+                    # Next, we look forward for as long as the triggers (resp. targets) are the
+                    # same. They all commute, so all are potential candidates for being simplified.
+                    common_B: Dict[int, int] = {}
+                    for k in range(j + 1, len(cnots)):
+                        if cnots[j][x] != cnots[k][x]:
+                            break
+                        # We take a note of the target (resp. trigger).
+                        common_B[cnots[k][y]] = k
+
+                    # Among all the candidates, find if they have a match.
+                    keys = common_A.keys() & common_B.keys()
+                    for key in keys:
+                        assert common_A[key] != common_B[key]
+                        # We perform the swap which removes the pivot.
+                        new_idx: List[int] = (
+                            [idx for idx in range(0, j) if idx != common_A[key]]
+                            + [common_B[key], common_A[key]]
+                            + [idx for idx in range(j + 1, len(cnots)) if idx != common_B[key]]
+                        )
+                        # Since we removed the pivot, the length should be one fewer.
+                        assert len(new_idx) == len(cnots) - 1
+                        cnots = [cnots[idx] for idx in new_idx]
+                        found_simplification = True
+                        break
+
+                    if found_simplification:
+                        break
+                if found_simplification:
+                    break
+
+            if found_simplification:
+                continue
+            break
+
+        return cnots
+
+    def _apply_cnots(prevh: Tuple[int, ...], currh: Tuple[int, ...]):
+        # This function applies in sequence the CNOTs from prevh and then currh. However, given
+        # that the h are sorted in Gray ordering and that some cancel each other, we can reduce
+        # the number of gates. See [4] for more details.
+
+        cnots: List[Tuple[int, int]] = []
+
+        if ladder_target:
+            cnots.extend((prevh[i], prevh[i + 1]) for i in reversed(range(len(prevh) - 1)))
+            cnots.extend((currh[i], currh[i + 1]) for i in range(len(currh) - 1))
+        else:
+            cnots.extend((prevh[i], prevh[-1]) for i in range(len(prevh) - 1))
+            cnots.extend((currh[i], currh[-1]) for i in range(len(currh) - 1))
+
+        cnots = _simplify_cnots(cnots)
+
+        for gate in (cirq.CNOT(qubits[c], qubits[t]) for c, t in cnots):
+            yield gate
+
+    previous_h: Tuple[int, ...] = ()
+    for h in sorted_hs:
+        w = combined.hamiltonians[h]
+
+        yield _apply_cnots(previous_h, h)
+
+        if len(h) >= 1:
+            yield cirq.Rz(rads=(theta * w)).on(qubits[h[-1]])
+
+        previous_h = h
+
+    # Flush the last CNOTs.
+    yield _apply_cnots(previous_h, ())
+
+
+@value.value_equality
+class HamiltonianGate(raw_types.Gate):
```

Comment:
I changed the name to your use your suggestion. I also took the liberty of renaming the file, since it seems to be better in line with the conventions.